Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -17,118 +17,320 @@ extension MongoCollection {
      */
     @discardableResult
     public func bulkWrite(_ requests: [WriteModel], options: BulkWriteOptions? = nil) throws -> BulkWriteResult? {
-        throw MongoError.commandError(message: "Unimplemented command")
+        if requests.isEmpty {
+            throw MongoError.invalidArgument(message: "requests cannot be empty")
+        }
+
+        let opts = try BsonEncoder().encode(options)
+        let bulk = mongoc_collection_create_bulk_operation_with_opts(self._collection, opts?.data)!
+        defer { mongoc_bulk_operation_destroy(bulk) }
+
+        var insertedIds = [Int: BsonValue?]()
+
+        for (index, model) in requests.enumerated() {
+            if let deleteOrUpdate = model as? DeleteOrUpdateWriteModel {
+                try deleteOrUpdate.addToBulkWrite(bulk: bulk)
+                continue
+            }
+
+            if let insert = model as? InsertOneModel<T> {
+                insertedIds[index] = try insert.addToBulkWrite(bulk: bulk)
+                continue
+            }
+
+            throw MongoError.invalidArgument(message: "Unsupported operation in requests[\(index)]: \(type(of: model))")
+        }
+
+        let reply = Document()
+        var error = bson_error_t()
+
+        if mongoc_bulk_operation_execute(bulk, reply.data, &error) == 0 {
```

Comment:
It returns the hint of the server used to execute the write (see: [Returns for `mongoc_bulk_operation_execute()`](http://mongoc.org/libmongoc/current/mongoc_bulk_operation_execute.html#returns)).

This isn't necessary with the query API, since [`mongoc_cursor_get_hint()`](http://mongoc.org/libmongoc/current/mongoc_cursor_get_hint.html) exists and allows that to be accessed after executing a query. There is technically a shortcoming with `command_with_opts()` and friends, but since those also allow a `serverId` option drivers that really want this information (i.e. PHP) can always do server selection themselves and rely on  `command_with_opts()` to bypass its own internal server selection.